### PR TITLE
Fix extent computation in QgsSpatiaLiteProvider

### PR DESCRIPTION
**Pull request by strk**
Trying to debug https://github.com/qgis/QGIS/pull/54646#discussion_r1355068330 I was lost in this function and thought I'd rewrite it.

### CODE REVIEW

File: src/providers/spatialite/qgsspatialiteprovider.cpp [hunk 1]
```diff
@@ -5820,52 +5820,54 @@ bool QgsSpatiaLiteProvider::getTableSummaryAbstractInterface( gaiaVectorLayerPtr
...
+
+  if ( rows != 1 )
   {
-    for ( i = 1; i <= rows; i++ )
-    {
-      QString count = results[( i * columns ) + 0];
-      mNumberFeatures = count.toLongLong();
+    QgsMessageLog::logMessage(
+      tr( "Spatialite: unexpected number of rows (%1) from aggregate query\nSQL: %1" )
+      .arg( rows ) .arg( sql)
```
Comment (lnicola) [inline src/providers/spatialite/qgsspatialiteprovider.cpp:5855]:
```suggestion
      .arg( rows ) .arg( sql )
```

File: src/providers/spatialite/qgsspatialiteprovider.cpp [hunk 2]
```diff
@@ -5820,52 +5820,54 @@ bool QgsSpatiaLiteProvider::getTableSummaryAbstractInterface( gaiaVectorLayerPtr
...
-      {
-        mLayerExtent.setMinimal();
-      }
-      else
-      {
-        QString minX = results[( i * columns ) + 1];
-        QString minY = results[( i * columns ) + 2];
-        QString maxX = results[( i * columns ) + 3];
-        QString maxY = results[( i * columns ) + 4];
+  QString count = results[columns + 0];
```
Comment (lnicola) [inline src/providers/spatialite/qgsspatialiteprovider.cpp:5862]:
```suggestion
  // the first row of results contains the column names, so we skip that
  QString count = results[columns + 0];
```

File: src/providers/spatialite/qgsspatialiteprovider.cpp [hunk 3]
```diff
@@ -5820,52 +5820,54 @@ bool QgsSpatiaLiteProvider::getTableSummaryAbstractInterface( gaiaVectorLayerPtr
...
     sql += " WHERE ( " + mSubsetString + ')';
   }
 
-  ret = sqlite3_get_table( sqliteHandle( ), sql.toUtf8().constData(), &results, &rows, &columns, &errMsg );
+  char **results = nullptr;
+  int rows;
+  int columns;
+  char *errMsg = nullptr;
+
+  int ret = sqlite3_get_table( sqliteHandle( ), sql.toUtf8().constData(), &results, &rows, &columns, &errMsg );
```
Comment (lnicola) [inline src/providers/spatialite/qgsspatialiteprovider.cpp:5844]: Should this be

```suggestion
  int ret = sqlite3_get_table( sqliteHandle( ), sql.toUtf8( ).constData( ), &results, &rows, &columns, &errMsg );
```

Comment (strk) [reply src/providers/spatialite/qgsspatialiteprovider.cpp:5844]: is there a policy about these parens ? I see no space used below, maybe I shoudl remove it from the first element too

**Comment by lnicola** [comment]
Looks like I had an unsubmitted comment for some reason. Looks fine, I guess, but I really think moving to the standard query API is a better idea. But even without that, this looks like an improvement and probably fixes the mentioned issue.

**Review by lnicola — APPROVED** [review]

**Comment by strk** [comment]
The basic fix here, I believe, is handling the NULL min/max values of extent when number of rows is zero.
In that case, without this change, you get a 0.0 values instead of NULL, so a 0,0,0,0 QgsRectangle instead of a Null rectangle.
Nobody notices until you stop considering 0,0,0,0 not-null ( as in #54646 )

**Comment by strk** [comment]
If the empty string check is not good (CI will tell, I guess?) we can consider using proper null checking. This may be useful in that case: https://stackoverflow.com/questions/8961457/how-to-check-a-value-in-a-sqlite-column-is-null-or-not-with-c-api

File: src/providers/spatialite/qgsspatialiteprovider.cpp [hunk 4]
```diff
@@ -5820,52 +5820,61 @@ bool QgsSpatiaLiteProvider::getTableSummaryAbstractInterface( gaiaVectorLayerPtr
...
-  QString sql = QStringLiteral( "SELECT Count(1)%1 FROM %2" )
-                .arg( mGeometryColumn.isEmpty() ? QString() : QStringLiteral( ",Min(MbrMinX(%1)),Min(MbrMinY(%1)),Max(MbrMaxX(%1)),Max(MbrMaxY(%1))" ).arg( QgsSqliteUtils::quotedIdentifier( mGeometryColumn ) ),
-                      mQuery );
+  QString sql = QStringLiteral( "SELECT Count(1)" );
+
+  if ( mGeometryColumn.isEmpty() )
+  {
+    sql += QStringLiteral(
+             ", Min(MbrMinX(%1)), Min(MbrMinY(%1)), Max(MbrMaxX(%1)), Max(MbrMaxY(%1))"
+           ).arg( QgsSqliteUtils::quotedIdentifier( mGeometryColumn ) );
```
Comment (nyalldawson) [inline src/providers/spatialite/qgsspatialiteprovider.cpp:5831]: This looks wrong -- it's adding mGeometryColumn dependent sql filter only when the geometry column is empty?

File: src/providers/spatialite/qgsspatialiteprovider.cpp [hunk 5]
```diff
@@ -5820,52 +5820,61 @@ bool QgsSpatiaLiteProvider::getTableSummaryAbstractInterface( gaiaVectorLayerPtr
 
 bool QgsSpatiaLiteProvider::getTableSummary()
 {
-  int ret;
-  int i;
-  char **results = nullptr;
-  int rows;
-  int columns;
-  char *errMsg = nullptr;
+  mLayerExtent.setMinimal();
```
Comment (nyalldawson) [inline src/providers/spatialite/qgsspatialiteprovider.cpp:5823]: setMinimal is rather confusing. Could we add "setToNull" instead and deprecate setMinimal?

Comment (strk) [reply src/providers/spatialite/qgsspatialiteprovider.cpp:5823]: I'm very happy to add a setNull() to QgsRectangle and even to deprecate setMinimal, but is this the right PR ? Or should I just merge this back into PR GH-54646 and handle everything there ?

File: src/providers/spatialite/qgsspatialiteprovider.cpp (hunk 4 above)
Comment (strk) [reply src/providers/spatialite/qgsspatialiteprovider.cpp:5831]: Oops, force pushing a fix

File: src/providers/spatialite/qgsspatialiteprovider.cpp (hunk 5 above)
Comment (nyalldawson) [reply src/providers/spatialite/qgsspatialiteprovider.cpp]: Let's do it in a separate PR

File: src/providers/spatialite/qgsspatialiteprovider.cpp [hunk 6]
```diff
@@ -5820,52 +5820,61 @@ bool QgsSpatiaLiteProvider::getTableSummaryAbstractInterface( gaiaVectorLayerPtr
...
-  else
+
+  if ( rows != 1 )
   {
-    for ( i = 1; i <= rows; i++ )
-    {
-      QString count = results[( i * columns ) + 0];
-      mNumberFeatures = count.toLongLong();
+    QgsMessageLog::logMessage(
+      tr( "Spatialite: unexpected number of rows (%1) from aggregate query\nSQL: %1" )
```
Comment (elpaso) [inline src/providers/spatialite/qgsspatialiteprovider.cpp:5856]:
```suggestion
      tr( "Spatialite: unexpected number of rows (%1) from aggregate query\nSQL: %2" )
```

File: src/providers/spatialite/qgsspatialiteprovider.cpp [hunk 7]
```diff
@@ -5820,52 +5820,61 @@ bool QgsSpatiaLiteProvider::getTableSummaryAbstractInterface( gaiaVectorLayerPtr
...
-        QString maxY = results[( i * columns ) + 4];
+  // the first row of results contains the column names, so we skip that
+  QString count = results[columns + 0];
+  mNumberFeatures = count.toLongLong();
 
-        mLayerExtent.set( minX.toDouble(), minY.toDouble(), maxX.toDouble(), maxY.toDouble() );
-      }
+  if ( mNumberFeatures && ! mGeometryColumn.isEmpty() ) do
+    {
+      QString minX = results[columns + 1]; if ( minX.isEmpty() ) break;
```
Comment (elpaso) [inline src/providers/spatialite/qgsspatialiteprovider.cpp:5870]:
```suggestion
      const QString minX = results[columns + 1]; if ( minX.isEmpty() ) break;
```
and vars below

Comment (strk) [reply src/providers/spatialite/qgsspatialiteprovider.cpp]: Thanks for the review, force pushed with your suggested changes

File: src/providers/spatialite/qgsspatialiteprovider.cpp (hunk 6 above)
Comment (strk) [reply src/providers/spatialite/qgsspatialiteprovider.cpp]: force pushed with this change

File: src/providers/spatialite/qgsspatialiteprovider.cpp (hunk 4 above)
Comment (strk) [reply src/providers/spatialite/qgsspatialiteprovider.cpp]: This was done

File: src/providers/spatialite/qgsspatialiteprovider.cpp (hunk 5 above)
Comment (strk) [reply src/providers/spatialite/qgsspatialiteprovider.cpp]: I've added setNull() and deprecated setMinimal() in GH-54646

**Review by lnicola — APPROVED** [review]

**Comment by strk** [comment]
@elpaso I've addressed all your suggestions, are we ready to go now ? I need one approval to proceed

File: src/providers/spatialite/qgsspatialiteprovider.cpp (hunk 5 above)
Comment (strk) [reply src/providers/spatialite/qgsspatialiteprovider.cpp]: Given the deprecation of QgsRectangle::setMinimal() introduced other problems I put it in its own PR here: https://github.com/qgis/QGIS/pull/54934

@nyalldawson if we can merge this in while CI-green that'd be great, then GH-54934, if legit at this stage :)

**Review by elpaso — APPROVED** [review]

**Pull request merged**